From 821fd20a010df2407b44490704ed31787b2224f7 Mon Sep 17 00:00:00 2001
From: Kamal Mostafa <kamal@canonical.com>
Date: Wed, 28 Aug 2013 12:57:38 -0700
Subject: [PATCH 1/2] drm/i915: i915.disable_pch_pwm overrides PCH_PWM_ENABLE
 quirk

BugLink: https://bugzilla.kernel.org/show_bug.cgi?id=47941

Some BIOS configurations of Dell XPS13 are adversely affected by e85843b
("drm/i915: quirk no PCH_PWM_ENABLE for Dell XPS13 backlight") so provide a
boot param to inhibit the quirk, or force it on.

i915.disable_pch_pwm can be set to
 -1: auto (default; allows the disabling of pch_pwm by dmi quirk table match)
  0: inhibits the disabling of pch_pwm (overrides dmi quirk table match)
  1: forces the disabling of pch_pwm

Signed-off-by: Kamal Mostafa <kamal@canonical.com>
Cc: <stable@vger.kernel.org> # v3.8+
---
 drivers/gpu/drm/i915/i915_drv.c      |  4 ++++
 drivers/gpu/drm/i915/i915_drv.h      |  1 +
 drivers/gpu/drm/i915/intel_display.c | 11 ++++++++---
 3 files changed, 13 insertions(+), 3 deletions(-)

diff -Nurp linux-3.10.10/drivers/gpu/drm/i915/i915_drv.c linux-3.10.10.0001/drivers/gpu/drm/i915/i915_drv.c
--- linux-3.10.10/drivers/gpu/drm/i915/i915_drv.c	2013-09-03 12:31:00.909789087 +0300
+++ linux-3.10.10.0001/drivers/gpu/drm/i915/i915_drv.c	2013-09-03 12:34:53.794769399 +0300
@@ -128,6 +128,10 @@ module_param_named(disable_power_well, i
 MODULE_PARM_DESC(disable_power_well,
 		 "Disable the power well when possible (default: false)");
 
+int i915_disable_pch_pwm __read_mostly = -1;
+module_param_named(disable_pch_pwm, i915_disable_pch_pwm, int, 0600);
+MODULE_PARM_DESC(disable_pch_pwm, "disable PCH_PWM (default: -1 (auto))");
+
 static struct drm_driver driver;
 extern int intel_agp_enabled;
 
diff -Nurp linux-3.10.10/drivers/gpu/drm/i915/i915_drv.h linux-3.10.10.0001/drivers/gpu/drm/i915/i915_drv.h
--- linux-3.10.10/drivers/gpu/drm/i915/i915_drv.h	2013-09-03 12:31:00.909789087 +0300
+++ linux-3.10.10.0001/drivers/gpu/drm/i915/i915_drv.h	2013-09-03 12:36:28.066785184 +0300
@@ -1436,6 +1436,7 @@ extern bool i915_enable_hangcheck __read
 extern int i915_enable_ppgtt __read_mostly;
 extern unsigned int i915_preliminary_hw_support __read_mostly;
 extern int i915_disable_power_well __read_mostly;
+extern int i915_disable_pch_pwm __read_mostly;
 
 extern int i915_suspend(struct drm_device *dev, pm_message_t state);
 extern int i915_resume(struct drm_device *dev);
diff -Nurp linux-3.10.10/drivers/gpu/drm/i915/intel_display.c linux-3.10.10.0001/drivers/gpu/drm/i915/intel_display.c
--- linux-3.10.10/drivers/gpu/drm/i915/intel_display.c	2013-09-03 12:31:00.919789300 +0300
+++ linux-3.10.10.0001/drivers/gpu/drm/i915/intel_display.c	2013-09-03 12:34:53.814769840 +0300
@@ -8925,9 +8925,8 @@ static void quirk_invert_brightness(stru
  */
 static void quirk_no_pcm_pwm_enable(struct drm_device *dev)
 {
-	struct drm_i915_private *dev_priv = dev->dev_private;
-	dev_priv->quirks |= QUIRK_NO_PCH_PWM_ENABLE;
-	DRM_INFO("applying no-PCH_PWM_ENABLE quirk\n");
+	if (i915_disable_pch_pwm < 0)
+		i915_disable_pch_pwm = 1;
 }
 
 struct intel_quirk {
@@ -9025,6 +9024,12 @@ static void intel_init_quirks(struct drm
 		if (dmi_check_system(*intel_dmi_quirks[i].dmi_id_list) != 0)
 			intel_dmi_quirks[i].hook(dev);
 	}
+
+	if (i915_disable_pch_pwm == 1) {
+		struct drm_i915_private *dev_priv = dev->dev_private;
+		dev_priv->quirks |= QUIRK_NO_PCH_PWM_ENABLE;
+		DRM_INFO("applying no-PCH_PWM_ENABLE quirk\n");
+	}
 }
 
 /* Disable the VGA plane that we never use */
